### ISSUE TITLE
{.hygienic.} template with {.dirty:some_outer_symbol.} escape hatch

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#9903]
/cc @mratsim @Araq 

```nim
proc bar_local1() = discard
when not defined(foo): var x_outer_maybe = 0
var x_local1 = 0

template fun*() {.hygienic.} = # maybe become new default in future release after deprecation period
  # same: rules aren't changed for these:
  proc bar_injected() = discard
  proc bar_not_injected() {.gensym.} = discard
  var x_not_injected = 0
  var x_injected {.inject.} = 0

  # new: these cannot access calling scope
  bar_local1()
  bar_injected()
  bar_not_injected()
  echo x_local1

  # new: escape hatch to allow dirty access to calling scope
  {.dirty:bar_outer1.}
  bar_outer1()

  when defined(foo): # these inject pragma can be conditional
    {.dirty:x_outer_maybe¸.}
  echo x_outer_maybe

  # `result` follows these rules, nothing special
  {.dirty:result.}
  result.add 10
```

## related
* https://github.com/nim-lang/Nim/issues/8208 is related but I prefer my proposal because #8208 has serious issues:
  * would break too much code (my proposal doesn't break any code)
  * didn't provide escape hatch in hygienic templates
  * it required using `{.dirty.}` when some symbols needed outer scope, which is a blunt tool (and would prevent access to any local scope symbol)

**Comment by Araq** [comment]
This feature has a poor motivation. If you want to check your templates against typos, instantiate them. Nim should protect against bugs that cause problems at runtime, Nim is not an excercise in masochism that answers the question "how many pragmas can you add to programming language and get away with it?"

**Comment by timotheecour** [comment]
> If you want to check your templates against typos, instantiate them

That doesn't help (unless your test suite is exhaustive, which is never the case)

It's not about compiler errors vs runtime errors, it's about reducing likelihood of subtle bugs and catching bugs early
* patterns like `when compiles(fun()) : dothis else: dothat` makes the distinction between compile vs runtime error moot
* environmental differences means that even if your code compiles on your platform, you could've introduced a bug on another platform (not test suite is exhaustive)

Here's yet another point (different from above 2 points) showing how even if template is instantiated, you get different results after a subtle change (3rd party library gets upgraded)
main.nim:
```nim
import mytemplates
proc fun()=echo "from main"
callFun()
```

mytemplates.nim:
```nim
import mythirdparty
template callFun*() = fun()
```

mythirdparty.nim:
```nim
when defined(case_upgraded):
  # after upgrade, this proc is added
  proc fun*()=echo "from mythirdparty"
```

**Comment by mratsim** [comment]
I usually use `{.dirty.}` when it gets unhandy to pass everything I need to capture as parameters to the template. For example in my [pure Nim matrix multiplication](https://github.com/numforge/laser/blob/723299ea439bd10ffae0fda0601d5b153892303f/laser/primitives/matrix_multiplication/gemm_ukernel_generator.nim#L71-L98), I capture the following:

- result: a NimNode
- epilogue_name: a generated function name:
- T and V: types
- nb_scalars: a const
- simd_store_unaligned, simd_mul, simd_load_unaligned, simd_add, simd_fma: proc names

**Comment by Araq** [comment]
> patterns like when compiles(fun()) : dothis else: dothat makes the distinction between compile vs runtime error moot

I disagree completely. The distinction is an objective fact and clear since Nim lacks 'eval'-like facilities.

> environmental differences means that even if your code compiles on your platform, you could've introduced a bug on another platform (no test suite is exhaustive, especially for cross platform issues)

If you don't compile and test your code on platform X, X is an unsupported platform. To pretend/assume otherwise is wrong.

> Here's yet another point ...
> after 3rd party lib upgrade, code uses local scope proc, printingfrom mythirdparty
no compiler error, just a subtle hard to find bug.

Come back when that actually really happened in the real world. Then I will start to listen. After 3 people reporting the same problem, I will start to take it serious.

Of course, you can also mitigate the problems by using more ``from x import y, z`` if you use third party packages and you can use fewer templates in your code. Or -- I dunno -- have a sound development process in place where upgrading third party dependencies is treated as a risk like it should, moreso if your testing is inadequate.

**Comment by timotheecour** [comment]
@mratsim 
> I usually use {.dirty.} when it gets unhandy to pass everything I need to capture as parameters to the template

I find `dirty` templates to be rarely useful in practice **except** maybe for the case of a privately defined template, because it's impossible to use locally defined symbols (eg helpers) inside it (AFAIK), since it's entirely instantiated in caller scope. That makes its use limited. Indeed, the `dirty` examples you provided in https://github.com/numforge/laser/blob/723299ea439bd10ffae0fda0601d5b153892303f/laser/primitives/matrix_multiplication/gemm_ukernel_generator.nim#L71-L98 are private, not public, and I'm only counting 5 examples in stdlib of public dirty templates.

Hygienic templates with `dirty` escape hatch would allow both scope sanity by default + flexibility to access calling scope when needed.
